src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -1075,78 +1079,106 @@ void WindowEmperor::_finalizeSessionPersistence() const
+    {
+        // If the app is running elevated, we create files with a mandatory
+        // integrity label (ML) of "High" (HI) for reading and writing (NR, NW).
+        wil::unique_hlocal_security_descriptor sd;
+        SECURITY_ATTRIBUTES sa{};
+        if (admin)
+        {
+            unsigned long cb;
+            THROW_IF_WIN32_BOOL_FALSE(ConvertStringSecurityDescriptorToSecurityDescriptorW(L"S:(ML;;NRNW;;;HI)", SDDL_REVISION_1, wil::out_param_ptr<PSECURITY_DESCRIPTOR*>(sd), &cb));
+            sa.nLength = sizeof(SECURITY_ATTRIBUTES);
+            sa.lpSecurityDescriptor = sd.get();
+        }
+
+        // Persist all terminal buffers to "buffer_{guid}.txt" files.
+        // We remember the filenames so that we can clean up old ones later.
+        for (const auto& w : _windows)
         {
-            for (const auto& windowLayout : layouts)
+            const auto panes = w->Logic().Panes();
+            for (const auto pane : panes)
             {
-                for (const auto& actionAndArgs : windowLayout.TabLayout())
+                try
                 {
-                    const auto args = actionAndArgs.Args();
-                    NewTerminalArgs terminalArgs{ nullptr };
-
-                    if (const auto tabArgs = args.try_as<NewTabArgs>())
+                    const auto term = pane.try_as<winrt::TerminalApp::ITerminalPaneContent>();
+                    if (!term)
                     {
-                        terminalArgs = tabArgs.ContentArgs().try_as<NewTerminalArgs>();
+                        continue;
                     }
-                    else if (const auto paneArgs = args.try_as<SplitPaneArgs>())
+                    const auto control = term.GetTermControl();
+                    if (!control)
                     {
-                        terminalArgs = paneArgs.ContentArgs().try_as<NewTerminalArgs>();
+                        continue;
                     }
+                    const auto connection = control.Connection();
+                    if (!connection)
+                    {
+                        continue;
+                    }
+                    const auto sessionId = connection.SessionId();
+                    if (sessionId == winrt::guid{})
+                    {
+                        continue;
+                    }
+
+                    auto filename = fmt::format(FMT_COMPILE(L"{}{}.txt"), filenamePrefix, sessionId);
+                    const auto path = settingsDirectory / filename;
 
-                    if (terminalArgs)
+                    if (wil::unique_hfile file{ CreateFileW(path.c_str(), GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_DELETE, &sa, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr) })
                     {
-                        sessionIds.emplace(terminalArgs.SessionId());
+                        control.PersistTo(reinterpret_cast<int64_t>(file.get()));
+                        bufferFilenames.emplace(std::move(filename));

```

Comment:
Shouldn't we only do this if `_app.Logic().Settings().GlobalSettings().FirstWindowPreference() == FirstWindowPreference::PersistedLayoutAndContent`?